src/js/base/translate-parse-tree.js

```diff
@@ -0,0 +1,1312 @@
+    var pos2 = function(p1, p2) { return combineSrcloc(fileName, p1, p2); };
+    function makeListTr(arr, start, end, onto, f) {
+      var ret = onto || makeEmpty();
+      start = start || 0;
+      end = end || arr.length;
+      f = f || tr;
+      for (var i = end - 1; i >= start; i--)
+        ret = makeLink(f(arr[i]), ret);
+      return ret;
+    }
+    function makeListComma(arr, start, end, f) {
+      var ret = makeEmpty();
+      start = start || 0;
+      end = end || arr.length;
+      f = f || tr;
+      for (var i = end - 1; i >= start; i -= 2)
+        ret = makeLink(f(arr[i]), ret);
+      return ret;
+    }
+    function makeList(arr, start, end, onto) {
+      var ret = onto || makeEmpty();
+      start = start || 0;
+      end = end || arr.length;
+      for (var i = end - 1; i >= start; i--)
+        ret = makeLink(arr[i], ret);
+      return ret;
+    }
+    function name(tok) {
+      if (tok.value === "_")
+        return makeNode('s-underscore', pos(tok.pos));
+      else
+        return makeNode('s-name', pos(tok.pos), makeString(tok.value));
+    }
+    function symbol(tok) {
+      return makeString(tok.value);
+    }
+    function string(tok) {
+      if (tok.value.substring(0, 3) === "```")
+        return makeString(tok.value.slice(3, -3).trim());
+      else
+        return makeString(tok.value.slice(1, -1));
+    }
+    function number(tok) { return makeNumberFromString(tok.value); }
+    const translators = {
+      'program': function(node) {
+        var prelude = tr(node.kids[0]);
+        var body = tr(node.kids[1]);
+        return makeNode('s-program',
+                    pos(node.pos), prelude.provide, prelude.provideTypes, prelude.imports, body);
+      },
+      'prelude': function(node) {
+        var provide;
+        var provideTypes;
+        var kids = node.kids.slice(0);
+        if (kids.length > 0 && kids[0].name === "provide-stmt") {
+          provide = tr(kids.shift());
+        } else {
+          provide = makeNode('s-provide-none', pos(node.pos));
+        }
+        if (kids.length > 0 && kids[0].name === "provide-types-stmt") {
+          provideTypes = tr(kids.shift());
+        } else {
+          provideTypes = makeNode('s-provide-types-none', pos(node.pos));
+        }
+        return {
+          provide : provide,
+          provideTypes : provideTypes,
+          imports : makeListTr(kids)
+        };
+      },
+      'provide-stmt': function(node) {
+        if (node.kids.length === 2) {
+          // (provide-stmt PROVIDE STAR)
+          return makeNode('s-provide-all', pos(node.pos));
+        } else {
+          // (provide-stmt PROVIDE stmt END)
+          return makeNode('s-provide', pos(node.pos), tr(node.kids[1]));
+        }
+      },
+      'provide-types-stmt': function(node) {
+        if (node.kids[1].name === "STAR") {
+          return makeNode('s-provide-types-all', pos(node.pos));
+        } else {
+          // will produce record-ann
```

Comment:
This was always a bit of a pun and therefore a hack.  Since this is the only reason you need `getRecordFields`, it might be worth rewriting this to not be so punny.

Reply:
No no -- I'm saying, instead of calling `tr` on the first child, directly, copy the logic from the translation of `record-ann` into here, but *don't* create an actual `s-record-ann` object. Just grab the fields and use them they way you need to.